File: Libro de JavaScript/18-proyecto-biblia-consola/00_explorar_bd.js
// =============================================================================
// ARCHIVO: 00_explorar_bd.js
// TEMA: Exploración de la base de datos SQLite de la Biblia RV60
// =============================================================================
// DEPENDENCIA: npm install better-sqlite3

// Importar el módulo de base de datos SQLite (sincrónico)
const Database = require('better-sqlite3');
const path = require('path');

// Ruta a la base de datos relativa a este archivo
const DB_PATH = path.join(__dirname, '..', '..', 'datos', 'biblia_rv60.sqlite3');

/**
 * Abre la conexión a la base de datos SQLite.
 * Lanza un error si el archivo no existe o no es accesible.
 * @returns {Database} Instancia de la base de datos
 */
function abrirBD() {
  return new Database(DB_PATH, { readonly: true });
}

/**
 * Muestra todas las tablas disponibles en la base de datos.
 * @param {Database} db - Instancia de la base de datos
 */
function mostrarTablas(db) {
  console.log('\n=== TABLAS EN LA BASE DE DATOS ===');
  const tablas = db
    .prepare("SELECT name FROM sqlite_master WHERE type='table' ORDER BY name")
    .all();

  tablas.forEach(t => console.log(`  • ${t.name}`));
}

/**
 * Muestra los conteos totales de libros y versículos.
 * @param {Database} db - Instancia de la base de datos
 */
function mostrarConteos(db) {
  console.log('\n=== CONTEOS GENERALES ===');

  const totalLibros = db.prepare('SELECT COUNT(*) AS total FROM books').get();
  const totalVersiculos = db.prepare('SELECT COUNT(*) AS total FROM verses').get();

  console.log(`  Libros:     ${totalLibros.total}`);
  console.log(`  Versículos: ${totalVersiculos.total}`);
}

/**
 * Muestra los primeros 5 libros del Antiguo Testamento (book_number 1-39).
 * @param {Database} db - Instancia de la base de datos
 */
function mostrarPrimerosDel_AT(db) {
  console.log('\n=== PRIMEROS 5 LIBROS DEL AT ===');
  const libros = db
    .prepare(
      'SELECT book_number, short_name, long_name FROM books WHERE book_number <= 39 ORDER BY book_number LIMIT 5'
    )
    .all();

  libros.forEach(l =>
    console.log(`  [${String(l.book_number).padStart(2, '0')}] ${l.short_name.padEnd(5)} → ${l.long_name}`)
  );
}

/**
 * Muestra los últimos 5 libros del Nuevo Testamento (book_number 40-66).
 * @param {Database} db - Instancia de la base de datos
 */
function mostrarUltimosDelNT(db) {
  console.log('\n=== ÚLTIMOS 5 LIBROS DEL NT ===');
  const libros = db
    .prepare(
      'SELECT book_number, short_name, long_name FROM books WHERE book_number >= 40 ORDER BY book_number DESC LIMIT 5'
    )
    .all();

  // Invertir para mostrar en orden ascendente
  libros.reverse().forEach(l =>
    console.log(`  [${String(l.book_number).padStart(2, '0')}] ${l.short_name.padEnd(5)} → ${l.long_name}`)
  );
}

/**
 * Muestra la estructura de columnas de cada tabla.
 * @param {Database} db - Instancia de la base de datos
 */
function mostrarEstructura(db) {
  console.log('\n=== ESTRUCTURA DE TABLAS ===');
  const tablas = ['books', 'verses'];

  tablas.forEach(tabla => {
    console.log(`\n  Tabla: ${tabla}`);
    const columnas = db.prepare(`PRAGMA table_info(${tabla})`).all();
    columnas.forEach(c =>
      console.log(`    - ${c.name.padEnd(15)} ${c.type}`)
    );
  });
}

// ─── Punto de entrada principal ──────────────────────────────────────────────
let db;
try {
  db = abrirBD();
  console.log(`✓ Base de datos abierta: ${DB_PATH}`);

  mostrarTablas(db);
  mostrarConteos(db);
  mostrarPrimerosDel_AT(db);
  mostrarUltimosDelNT(db);
  mostrarEstructura(db);

  console.log('\n✓ Exploración completada.\n');
} catch (err) {
  console.error('\n✗ Error al abrir la base de datos:', err.message);
  console.error('  Verifica que el archivo exista en:', DB_PATH);
  process.exit(1);
} finally {
  // Cerrar siempre la conexión, sin importar si hubo error
  if (db) db.close();
}

File: Libro de Flutter/28-proyecto-biblia/00_explorar_bd.dart
// ============================================================
// MÓDULO 28 — Proyecto Biblia (Dart consola)
// Archivo: 00_explorar_bd.dart
//
// DEPENDENCIA: sqlite3: ^2.x.x en pubspec.yaml
//              path: ^1.9.0 en pubspec.yaml
//
// Propósito: explorar la estructura de la base de datos
// biblia_rv60.sqlite3 y verificar que está correctamente
// formada antes de operar sobre ella.
//
// Ejecutar con:
//   dart run lib/00_explorar_bd.dart
// ============================================================

import 'dart:io';
import 'package:path/path.dart' as path;
import 'package:sqlite3/sqlite3.dart';

void main() async {
  // Calcular la ruta absoluta a la BD de forma portátil.
  // Platform.script devuelve la URI del script en ejecución.
  final String scriptDir =
      path.dirname(Platform.script.toFilePath());
  final String dbPath =
      path.join(scriptDir, '..', '..', 'datos', 'biblia_rv60.sqlite3');
  final String rutaAbsoluta = path.normalize(dbPath);

  print('=' * 55);
  print('  EXPLORADOR — Biblia Reina-Valera 1960');
  print('=' * 55);
  print('Ruta de la BD: $rutaAbsoluta\n');

  // Verificar que el archivo existe antes de intentar abrirlo.
  if (!File(rutaAbsoluta).existsSync()) {
    print('ERROR: No se encontró el archivo de base de datos.');
    print('Verifica que biblia_rv60.sqlite3 esté en:');
    print('  ${path.dirname(rutaAbsoluta)}');
    exit(1);
  }

  // Abrir la conexión. sqlite3.open() lanza SqliteException si hay error.
  late final Database db;
  try {
    db = sqlite3.open(rutaAbsoluta, mode: OpenMode.readOnly);
    print('Conexión exitosa a la base de datos.\n');
  } on SqliteException catch (e) {
    print('ERROR al abrir la BD: ${e.message}');
    exit(1);
  }

  try {
    _mostrarTablas(db);
    _mostrarConteos(db);
    _mostrarPrimerosCincoLibros(db);
    _mostrarUltimosCincoLibros(db);
    _mostrarInfoVersion(db);
    _mostrarEjemploVersiculo(db);
  } finally {
    // Siempre cerrar la conexión, incluso si hay error.
    db.dispose();
    print('\nConexión cerrada.');
  }
}

/// Muestra las tablas disponibles en la base de datos.
void _mostrarTablas(Database db) {
  print('── TABLAS EN LA BASE DE DATOS ──────────────────────');
  final resultado = db.select(
    "SELECT name, type FROM sqlite_master "
    "WHERE type IN ('table', 'view') ORDER BY name",
  );
  for (final fila in resultado) {
    print('  ${fila['type'].toString().toUpperCase().padRight(6)} → ${fila['name']}');
  }
  print('');
}

/// Muestra el conteo total de libros y versículos.
void _mostrarConteos(Database db) {
  print('── CONTEOS GENERALES ───────────────────────────────');

  final libros = db.select('SELECT COUNT(*) AS total FROM books');
  final versiculos = db.select('SELECT COUNT(*) AS total FROM verses');
  final capitulos = db.select(
    'SELECT COUNT(DISTINCT book_number || "-" || chapter) AS total FROM verses',
  );

  print('  Libros:     ${libros.first['total']}');
  print('  Capítulos:  ${capitulos.first['total']}');
  print('  Versículos: ${versiculos.first['total']}');
  print('');
}

/// Muestra los primeros 5 libros del Antiguo Testamento.
void _mostrarPrimerosCincoLibros(Database db) {
  print('── PRIMEROS 5 LIBROS (AT) ──────────────────────────');
  final resultado = db.select(
    'SELECT book_number, short_name, long_name FROM books '
    'ORDER BY book_number ASC LIMIT 5',
  );
  _imprimirTablaLibros(resultado);
  print('');
}

/// Muestra los últimos 5 libros del Nuevo Testamento.
void _mostrarUltimosCincoLibros(Database db) {
  print('── ÚLTIMOS 5 LIBROS (NT) ───────────────────────────');
  final resultado = db.select(
    'SELECT book_number, short_name, long_name FROM books '
    'ORDER BY book_number DESC LIMIT 5',
  );
  // Invertir para mostrar en orden ascendente
  final listaInvertida = resultado.reversed.toList();
  _imprimirTablaLibros(listaInvertida);
  print('');
}

/// Formatea e imprime una lista de filas de libros en tabla.
void _imprimirTablaLibros(List<Row> filas) {
  print('  N°   Abrev   Nombre');
  print('  ───  ──────  ──────────────────────');
  for (final fila in filas) {
    final num = fila['book_number'].toString().padLeft(3);
    final abrev = fila['short_name'].toString().padRight(6);
    final nombre = fila['long_name'].toString();
    print('  $num  $abrev  $nombre');
  }
}

/// Muestra las primeras 3 filas de la tabla info con metadatos de la versión.
void _mostrarInfoVersion(Database db) {
  print('── INFORMACIÓN DE LA VERSIÓN (tabla info) ──────────');
  final resultado = db.select('SELECT name, value FROM info LIMIT 3');
  for (final fila in resultado) {
    final nombre = fila['name'].toString().padRight(16);
    print('  $nombre → ${fila['value']}');
  }
  print('');
}

/// Muestra Juan 3:16 como ejemplo de versículo con marcas Strong.
void _mostrarEjemploVersiculo(Database db) {
  print('── EJEMPLO DE VERSÍCULO CON MARCAS STRONG ──────────');

  // Juan = libro 43 en la numeración estándar de esta BD
  final resultado = db.select(
    'SELECT v.text, b.long_name '
    'FROM verses v JOIN books b ON v.book_number = b.book_number '
    'WHERE v.book_number = 43 AND v.chapter = 3 AND v.verse = 16',
  );

  if (resultado.isEmpty) {
    print('  No se encontró el versículo.');
    return;
  }

  final textoOriginal = resultado.first['text'].toString();
  final textoLimpio = textoOriginal.replaceAll(RegExp(r'<[^>]+>'), '');
  final libro = resultado.first['long_name'].toString();

  print('  Referencia: $libro 3:16');
  print('  Con marcas: $textoOriginal');
  print('  Limpio:     $textoLimpio');
}
